### ISSUE TITLE
Interpretation: What does max measure in Timer and DistributionSummary

### ISSUE THREAD

**Issue by tarunrathor-pro** [micrometer-metrics/micrometer-docs#132]
Based on available documentation it is not clear what does the following micrometer/promethues  measure.
http_server_requests_seconds_max

**Comment by aprevost** [comment]
I'd just like to second this request. We were trying to determine exactly what the _max metric generated by a Timer represented, and what time period it was calculated over, and couldn't find the answer anywhere until we found this issue.

Although one of the first places I looked was in the micrometer documentation section on Timers that you referred me to above, it wasn't at all clear to me that the NOTE referencing TimeWindowMax that the replies above was (in part) an answer to our questions.

I think it should be much more explicitly documented that for any Timer, separate metrics with the _count, _sum, and _max suffixes (for Prometheus, at least) will be generated. And what the default time window is over which the _max value calculated (somebody above said 1 minute?), and how to change that time window, if we wish to do so.

**Comment by DanielYWoo** [comment]
> Time window size will be the step size of the meter registry unless expiry in DistributionStatisticConfig is set to other value explicitly

Prometheus requires the bucket values to be accumulative and never decay, so I must set the expiry to null for Prometheus. However this `_max` is a gauge calculated on a time window, and it must expire and can go down, there must be an expiry for it. So, does it mean the `_max` will never go down for the Prometheus?

**Comment by jonatan-ivanov** [comment]
>Prometheus requires the bucket values to be accumulative and never decay

Buckets in a Prometheus Histogram are Counters so they are monotonic (never decay). But Micrometer's Max is a Gauge not a Histogram, so monotonicity this does not apply, a Gauge can go up or down freely.

>so I must set the expiry to null for Prometheus.

I don't think so, see above.. Also, a never expiring max is not very useful in general.

>So, does it mean the _max will never go down for the Prometheus?

No, max will decay, see above.

>Is this expiry configuration only for "push-based" endpoints, not for "pull-based" like Prometheus? If so, how is the _max gauge calculated when it is scraped from Promethues?

This is for most of the registries including Prometheususing, see: [`TimeWindowMax`](https://github.com/micrometer-metrics/micrometer/blob/973581c0839435ad0dcb5a8731d645f21c121f23/micrometer-core/src/main/java/io/micrometer/core/instrument/distribution/TimeWindowMax.java).

**Comment by DanielYWoo** [comment]
> Max for basic Timer implementations such as CumulativeTimer, StepTimer is a time window max (TimeWindowMax). It means that its value is the maximum value during a time window. If no new values are recorded for the time window length, the max will be reset to 0 as a new time window starts. Time window size will be the step size of the meter registry unless expiry in DistributionStatisticConfig is set to other value explicitly.

Let's say I have a Timer
```
        Timer.builder("chassis.redis1.latency") // Note, "." will be converted to "_" for Prometheus format
               .maximumExpectedValue(Duration.ofMillis(1000))
               .publishPercentileHistogram()
               .register(meterRegistry).record(someDuration, TimeUnit.MILLISECONDS);
```

It typically generates a max gauge, and some counters (_sum, _count and the buckets). e..g.,
```
# HELP chassis_redis1_latency_seconds_max
# TYPE chassis_redis1_latency_seconds_max gauge
chassis_redis1_latency_seconds_max{app="demo",node="001",service_name="demo",} 1.133
# HELP chassis_redis1_latency_seconds
# TYPE chassis_redis1_latency_seconds histogram
chassis_redis1_latency_seconds_count{app="demo",node="001",service_name="demo",} 200.0
chassis_redis1_latency_seconds_sum{app="demo",node="001",service_name="demo",} 204.375
chassis_redis1_latency_seconds_bucket{app="demo",node="001",service_name="demo",le="0.001",} 0.0
chassis_redis1_latency_seconds_bucket{app="demo",node="001",service_name="demo",le="0.001048576",} 0.0
chassis_redis1_latency_seconds_bucket{app="demo",node="001",service_name="demo",le="0.001398101",} 0.0
...
```

From the doc, `Time window size will be the step size of the meter registry unless expiry in DistributionStatisticConfig is set to other value explicitly.` it looks like the `expiry` controls both the max gauge and the counters. I need to set the expiry for the max gauge as you said it is useless if no decay, but it will make counters decay as well, which is a MUST-NOT for Prometheus. There should be two independent values `expiryMaxGauge` and `expiryCounters` to control the decay, or we simply ignore the `expiry` in the PrometheusMeterRegistry. What is the expected behavior here?

**Comment by jonatan-ivanov** [comment]
>it looks like the expiry controls both the max gauge and the counters

To me I don't think it does. The doc in your quote mentions the "step" size, it is talking about "step" registries. Prometheus is "cumulative" so the part you quoted should not apply to it. Maybe we should somehow call this out in the docs. Please let us know if you have any suggestions.

>I need to set the expiry for the max gauge as you said it is useless if no decay

You don't need to set anything to make it decay, max decays by default. Leaving buffer length and expiry on their defaults is also the recommended way.

> but it will make counters decay as well

Did you try this? Prometheus has its own histogram (`PrometheusHistogram`) where we do this: `.expiry(Duration.ofDays(1825)) // effectively never rolls over`. So if you see decaying behavior on counters that must be a bug. In that case, could you please open a new issue with a minimal Java sample that reproduces the issue so that we can troubleshoot?

>What is the expected behavior for Prometheus here?

Max should decay, counters should not, please the previous section and let us know if you can reproduce a behavior where this is not the case.

**Comment by DanielYWoo** [comment]
> Did you try this? Prometheus has its own histogram (PrometheusHistogram) where we do this: .expiry(Duration.ofDays(1825)) // effectively never rolls over. So if you see decaying behavior on counters that must be a bug. In that case, could you please open a new issue with a minimal Java sample that reproduces the issue so that we can troubleshoot?

Not tested, but I read the source code and I guess the same. That's why I was confused that the Prometheus implementation is not like the docs says.

> The doc in your quote mentions the "step" size, it is talking about "step" registries. Prometheus is "cumulative" so the part you quoted should not apply to it. Maybe we should somehow call this out in the docs. Please let us know if you have any suggestions.

So, a Timer generates a gauge and many counters: 
1. the gaugae `_max` will respect the `expiry` parameter and always has a decay time window
2. the counters: `_sum`, `_count` and the buckets depends
2.1 they do not respect `expiry` with cumulative collectors like `PrometheusMeterRegistry` (pull-based).
2.2 they respect `expiry` with step-based collectors like `StepMeterRegistry` (push-based).

Really appreciate for the clarification. I hope we can change the doc just a little bit, when I read the doc I was confused about what the "step" is in Promethues. If my comment above is correct, do you mind if I create a PR about the doc?

**Comment by jonatan-ivanov** [comment]
Please do not guess based on the code but test it. :)

What you wrote above seems right just one more thing: there are registries that do not support histograms at all and there are registries that do support cumulative and delta too.

If you have suggestions for the docs, please feel free to file a PR.

**Comment by DanielYWoo** [comment]
I have created a PR, pls review:  https://github.com/micrometer-metrics/micrometer-docs/pull/333
@jonatan-ivanov

**Comment by jonatan-ivanov** [comment]
Closed by #333
